### PR TITLE
Remove state out of web servers - remove Project cache

### PR DESCRIPTION
**Pull request by jamiesjc**
Currently, we are using "projectsById" and "projectsByName" to keep the cache info about projects. In order to scale out web servers and remove the state, we need to query DB directly for the project info instead of referring to the cache.

### CODE REVIEW

File: azkaban-common/src/main/java/azkaban/project/ProjectManager.java [hunk 1]
```diff
@@ -175,7 +150,16 @@ public Props getProps() {
   }
 
   public List<Project> getProjects() {
-    return new ArrayList<Project>(projectsById.values());
+    List<Project> projects;
```
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/project/ProjectManager.java]: initialize projects

File: azkaban-common/src/main/java/azkaban/project/ProjectManager.java [hunk 2]
```diff
@@ -159,8 +134,8 @@ public Props getProps() {
       logger.error("Bad regex pattern " + regexPattern);
       return array;
     }
-
-    for (Project project : projectsById.values()) {
+    List<Project> projects = getProjects();
+    for (Project project : projects) {
```
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/project/ProjectManager.java]: for (Project project : getProjects()), no need to introduce another variable

File: azkaban-common/src/main/java/azkaban/project/ProjectManager.java [hunk 3]
```diff
@@ -142,7 +116,8 @@ public Props getProps() {
 
   public List<Project> getGroupProjects(User user) {
     List<Project> array = new ArrayList<Project>();
```
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/project/ProjectManager.java]: can you make the name more explicit? array doesn't indicate anything.

File: azkaban-common/src/main/java/azkaban/project/ProjectManager.java [hunk 4]
```diff
@@ -187,7 +171,8 @@ public Props getProps() {
       logger.error("Bad regex pattern " + regexPattern);
       return allProjects;
     }
-    for (Project project : getProjects()) {
+    List<Project> projects = getProjects();
+    for (Project project : projects) {
```
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/project/ProjectManager.java]: why do we need to introduce projects here? why not just for (Project project : getProjects())

File: azkaban-common/src/main/java/azkaban/project/ProjectManager.java [hunk 5]
```diff
@@ -196,21 +181,13 @@ public Props getProps() {
...
-
-    /**
      * Checks if a project is active using project_id
      *
      * @param name
      */
     public Boolean isActiveProject(int id) {
-        return projectsById.containsKey(id);
+      Project project = getProject(id);
+      return project != null;
```
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/project/ProjectManager.java]: getProject(id) != null simpler

File: azkaban-common/src/main/java/azkaban/project/ProjectManager.java [hunk 6]
```diff
@@ -222,14 +199,11 @@ public Boolean isActiveProject(int id) {
...
-            try {
-                fetchedProject = projectLoader.fetchProjectByName(name);
-            } catch (ProjectManagerException e) {
-                logger.error("Could not load project from store.", e);
-            }
+        try {
+            fetchedProject = projectLoader.fetchProjectByName(name);
+            loadAllProjectFlows(fetchedProject);
+        } catch (ProjectManagerException e) {
+            logger.error("Could not load project from store.", e);
```
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/project/ProjectManager.java]: log the project name as well.

**Comment by kunkun-tang** [comment]
Just double check Is this a complete RB to remove project info out of cache? I thought we probably need to modify **class project** as well

File: azkaban-common/src/main/java/azkaban/project/ProjectManager.java [hunk 7]
```diff
@@ -196,21 +181,13 @@ public Props getProps() {
...
-    public Boolean isActiveProject(String name) {
-        return projectsByName.containsKey(name);
-    }
-
-    /**
      * Checks if a project is active using project_id
      *
      * @param name
      */
     public Boolean isActiveProject(int id) {
```
Comment (kunkun-tang) [inline azkaban-common/src/main/java/azkaban/project/ProjectManager.java]: Looks like this method is not used any longer? Who will call it ?Should we just remove it? Besides, the method name sounds a bit confused here. We should detect exist here, rather than active, right?

File: azkaban-common/src/main/java/azkaban/project/ProjectManager.java (hunk 6 above)
Comment (kunkun-tang) [inline azkaban-common/src/main/java/azkaban/project/ProjectManager.java]: +1

File: azkaban-common/src/main/java/azkaban/project/ProjectManager.java [hunk 8]
```diff
@@ -118,17 +95,14 @@ private void loadAllProjectFlows(Project project) {
...
-
   public Props getProps() {
     return props;
   }
 
   public List<Project> getUserProjects(User user) {
     ArrayList<Project> array = new ArrayList<Project>();
-    for (Project project : projectsById.values()) {
+    List<Project> projects = getProjects();
+    for (Project project : projects) {
```
Comment (kunkun-tang) [inline azkaban-common/src/main/java/azkaban/project/ProjectManager.java]: Should be safe to combine.

File: azkaban-common/src/main/java/azkaban/project/ProjectManager.java [hunk 9]
```diff
@@ -243,14 +217,11 @@ public Project getProject(String name) {
...
-            fetchedProject = projectsById.get(id);
-        } else {
-            try {
-                fetchedProject = projectLoader.fetchProjectById(id);
-            } catch (ProjectManagerException e) {
-                logger.error("Could not load project from store.", e);
-            }
+        try {
+            fetchedProject = projectLoader.fetchProjectById(id);
+            loadAllProjectFlows(fetchedProject);
```
Comment (suvodeep-pyne) [inline azkaban-common/src/main/java/azkaban/project/ProjectManager.java]: What does load all project flows do? Are there cases where I need the project and not load the flows?

If they need to always be done together. they should be encapsulated in a method probably in `ProjectLoader` if appropriate.

File: azkaban-common/src/main/java/azkaban/project/ProjectManager.java (hunk 1 above)
Comment (suvodeep-pyne) [inline azkaban-common/src/main/java/azkaban/project/ProjectManager.java]: This may not be required since you either populate the project or you throw a runtime exception.

File: azkaban-common/src/main/java/azkaban/project/ProjectManager.java (hunk 7 above)
Comment (jamiesjc) [inline azkaban-common/src/main/java/azkaban/project/ProjectManager.java]: It is called in handlePurgeProject. A project is active means the boolean active is set to true in DB for that project. I'll add some comments about it.

File: azkaban-common/src/main/java/azkaban/project/ProjectManager.java (hunk 9 above)
Comment (jamiesjc) [inline azkaban-common/src/main/java/azkaban/project/ProjectManager.java]: loadAllProjectFlows() will set the Map<string, Flow> flows for the project. getProject() is called in so many places, not able to enumerate them all...Most of them (>90%) will need to get the flow info, probably not all.

**Comment by jamiesjc** [comment]
@chengren311 Project table has below data: | id | name                               | active | modified_time | create_time   | version | last_modified_by | description                          | enc_type | settings_blob. 
Moving cache info from web servers to DB might downgrade the performance a little bit. Currently the DB query takes about 0.02 sec. Consider our current scale and load, this shouldn't affect much. And we will have master-slave DB plan in the future to overcome potential issues.

**Comment by jamiesjc** [comment]
@suvodeep-pyne 
The results of these methods are not cached. I'm planning to add some metrics to count how often these methods are called.

File: azkaban-common/src/main/java/azkaban/project/ProjectManager.java [hunk 10]
```diff
@@ -118,64 +95,68 @@ private void loadAllProjectFlows(Project project) {
...
-    return array;
+    return userProjects;
   }
 
   public List<Project> getProjects() {
-    return new ArrayList<Project>(projectsById.values());
+    List<Project> projects;
+    try {
+      projects = projectLoader.fetchAllActiveProjects();
+      for (Project proj : projects) {
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/project/ProjectManager.java]: This is a more expensive operation. 

Is this information always needed when getProjects() is called?

File: azkaban-common/src/main/java/azkaban/project/ProjectManager.java [hunk 11]
```diff
@@ -196,61 +177,44 @@ public Props getProps() {
...
-            try {
-                fetchedProject = projectLoader.fetchProjectById(id);
-            } catch (ProjectManagerException e) {
-                logger.error("Could not load project from store.", e);
-            }
+        try {
+            fetchedProject = projectLoader.fetchProjectById(id);
+            loadAllProjectFlows(fetchedProject);
+        } catch (ProjectManagerException e) {
+            logger.error("Could not load project from store.", e);
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/project/ProjectManager.java]: Add the project id to the log?

File: azkaban-common/src/main/java/azkaban/project/ProjectManager.java [hunk 12]
```diff
@@ -268,16 +232,10 @@ public Project createProject(String projectName, String description,
           "Project names must start with a letter, followed by any number of letters, digits, '-' or '_'.");
     }
 
-    if (projectsByName.containsKey(projectName)) {
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/project/ProjectManager.java]: If the project already exists, a DB exception will be thrown now. Right? How is that new exception handled differently from the existing one?

**Comment by HappyRay** [comment]
Could you describe how you test this change? I can't find existing unit tests for this class.

File: azkaban-common/src/test/java/azkaban/project/MockProjectLoader.java [hunk 13]
```diff
@@ -37,30 +39,52 @@ public MockProjectLoader(File dir) {
...
+      }
+    }
+    return activeProjects;
   }
 
   @Override
   public Project fetchProjectById(int id) throws ProjectManagerException {
-    // TODO Auto-generated method stub
-    return null;
+    System.out.println("MockProjectLoader: fetch project by id " + id);
```
Comment (ameyamk) [inline azkaban-common/src/test/java/azkaban/project/MockProjectLoader.java]: Any reason why we use System.out instead of logger?

File: azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java [hunk 14]
```diff
@@ -193,17 +193,23 @@ private Project fetchProjectByName(Connection connection, String name)
...
-        Project project = null;
+        Project project;
         ProjectResultHandler handler = new ProjectResultHandler();
+        // select active project from db first, if not exist, select inactive one.
+        // At most one active project with the same name exists in db.
         try {
             List<Project> projects =
                 runner.query(connection,
-                    ProjectResultHandler.SELECT_PROJECT_BY_NAME, handler, name);
+                      ProjectResultHandler.SELECT_ACTIVE_PROJECT_BY_NAME, handler, name);
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java]: Looks like this is an existing query. How was it used?

Is there a need to return inactive project?

File: azkaban-common/src/main/java/azkaban/project/ProjectManager.java (hunk 1 above)
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/project/ProjectManager.java]: I think it is a good practice to always initialize local variables.

File: azkaban-common/src/main/java/azkaban/project/ProjectManager.java [hunk 15]
```diff
@@ -196,61 +174,44 @@ public Props getProps() {
...
-        return projectsByName.containsKey(name);
-    }
-
-    /**
      * Checks if a project is active using project_id
      *
-     * @param name
+     * @param id
      */
     public Boolean isActiveProject(int id) {
```
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/project/ProjectManager.java]: How was this method used?

File: azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+import org.junit.Test;
+
+public class ProjectManagerTest {
+
+  private File workingDir;
+
+  @Before
+  public void setUp() throws Exception {
+    System.out.println("projectManagerTest: create temp dir");
+    synchronized (this) {
```
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java]: Why does it need to be synchronized?

File: azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+      workingDir = new File("_AzkabanTestDir_" + System.currentTimeMillis());
```
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java]: Is this helpful? http://junit.org/junit4/javadoc/4.12/org/junit/rules/TemporaryFolder.html

File: azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+    Project project1 = createProjectUtil(manager, projectName1, projectDescription1, user);
+
+    String projectName2 = "myTestFetchAllProjects_2";
+    String projectDescription2 = "This is to test fetching all projects: project_2";
+    Project project2 = createProjectUtil(manager, projectName2, projectDescription2, user);
+
+    List<Project> fetchedProjects = manager.getProjects();
+    Assert.assertTrue("Fetched project_1", fetchedProjects.contains(project1));
+    Assert.assertTrue("Fetched project_2", fetchedProjects.contains(project2));
+    System.out.println("Fetched all projects successfully");
```
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java]: Is this log necessary?

**Comment by HappyRay** [comment]
What's the rollback strategy if something goes wrong?

**Comment by HappyRay** [comment]
Are there places where later calls can use the previous DB call results in the context of the same user request? How does it affect consistency?
e.g. one user request translates to  get all projects, doing something else, get one project.

**Comment by HappyRay** [comment]
If we refactor project class and separate out the mutable states from immutable states, can we cache project information?

File: azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+
+    Assert.assertTrue("Project Id set", project.getId() > -1);
+    Assert.assertEquals("Project name", projectName, project.getName());
+    Assert.assertEquals("Project description", description,
+        project.getDescription());
+    Assert.assertTrue("Project is active", project.isActive());
+    return project;
+  }
+
+  private void assertProjectMemberEquals(Project p1, Project p2) {
```
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java]: Can this method be reused? 
azkaban.project.JdbcProjectLoaderTest#assertProjectMemberEquals

Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java]: Also have you consider overwriting the equals method of the Project class?

**Comment by HappyRay** [comment]
Removing the project related cache is not enough to scale out web servers. e.g.
azkaban.project.JdbcProjectLoader#createNewProject(java.sql.Connection, java.lang.String, java.lang.String, azkaban.user.User) there will be a race condition if two users try to create projects with the same name at the same time.
Currently the name column is not marked unique, it may work if we change it to unique and let one of the insert fail.

**Comment by HappyRay** [comment]
I don't see anywhere in the non test code that azkaban.project.Project#isActive is used.

File: azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java (hunk 19 above)
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java]: see azkaban.project.Project#equals
What's the source member used for? It doesn't seem to be used.
azkaban.project.Project#source

File: azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+  }
+
+  @Test
+  public void testCreateProject() throws ProjectManagerException {
+    System.out.println("projectManagerTest: testCreateProject");
+    ProjectManager manager = createProjectManager();
+    String projectName = "myTestCreateProject";
+    String projectDescription = "This is to test creating a project";
+    User user = new User("testUser");
+    createProjectUtil(manager, projectName, projectDescription, user);
```
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java]: It wasn't clear to me at first where the assertions are.
Consider keeping the util function just as a helper to create projects but not do assertions.

File: azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+      }
+    }
+  }
+
+  @Test
+  public void testCreateProject() throws ProjectManagerException {
+    System.out.println("projectManagerTest: testCreateProject");
+    ProjectManager manager = createProjectManager();
+    String projectName = "myTestCreateProject";
+    String projectDescription = "This is to test creating a project";
```
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java]: Remove some code duplications?

File: azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+    User user = new User("testUser");
+    createProjectUtil(manager, projectName, projectDescription, user);
+    System.out.println("Created project successfully");
+  }
+
+  @Test
+  public void testRemoveProject() throws ProjectManagerException {
```
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java]: Since the method is simple
  public synchronized Project removeProject(Project project, User deleter)
      throws ProjectManagerException {
    projectLoader.removeProject(project, deleter.getUserId());
    projectLoader.postEvent(project, EventType.DELETED, deleter.getUserId(),
        null);
    return project;
  }
Would using mockito to verify the internal calls be a better test? This will cover for example the user parameter is passed correctly.

Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java]: Do you plan to update the unit tests for JdbcProjectLoader too? It's ok to do that in a separate pull request.

File: azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java [hunk 23]
```diff
@@ -0,0 +1,162 @@
...
+  @Test
+  public void testUploadProject() throws ProjectManagerException {
+    System.out.println("projectManagerTest: testUploadProject");
+    ProjectManager manager = createProjectManager();
+    String projectName = "myTestUploadProject";
+    String projectDescription = "This is to test uploading a project";
+    User user = new User("testUser");
+    Project project = createProjectUtil(manager, projectName, projectDescription, user);
+
+    File testDir = new File("src/test/resources/project/testjob/testjob.zip");
```
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java]: Doesn't this code depend on the working directory to be set in certain way?

see http://stackoverflow.com/questions/28673651/how-to-get-the-path-of-src-test-resources-directory-in-junit for an alternative. You can find existing tests that use this technique.

File: azkaban-common/src/test/java/azkaban/project/MockProjectLoader.java (hunk 13 above)
Comment (jamiesjc) [inline azkaban-common/src/test/java/azkaban/project/MockProjectLoader.java]: The output of System.out will be redirected to a local file. This is already implemented for all other unit tests.

File: azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java [hunk 24]
```diff
@@ -193,17 +193,23 @@ private Project fetchProjectByName(Connection connection, String name)
         throws ProjectManagerException {
         QueryRunner runner = new QueryRunner();
         // Fetch the project
-        Project project = null;
+        Project project;
```
Comment (jamiesjc) [inline azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java]: Intellij gives this warning saying variable 'project' initializer null is redundant.  It will be assigned value later in project = projects.get(0); It there is any exception, then we don't even need to return project.

File: azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java (hunk 14 above)
Comment (jamiesjc) [inline azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java]: SELECT_ACTIVE_PROJECT_BY_NAME is used when creating a new project.
I also doubt whether we really need to return inactive project. In the future we can probably change the logic to remove inactive projects totally and only return active one. But for now, I just keep the existing logic in case something is breaking.

File: azkaban-common/src/main/java/azkaban/project/ProjectManager.java (hunk 1 above)
Comment (jamiesjc) [inline azkaban-common/src/main/java/azkaban/project/ProjectManager.java]: Intellij will give some optimization suggestions like the initialization is redundant and should be removed.

File: azkaban-common/src/main/java/azkaban/project/ProjectManager.java (hunk 15 above)
Comment (jamiesjc) [inline azkaban-common/src/main/java/azkaban/project/ProjectManager.java]: It is used in handlePurgeProject() which will do the actual cleanup of the projects. If the project is still active, then it will return some error saying the project should be deleted before purging.

File: azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java (hunk 16 above)
Comment (jamiesjc) [inline azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java]: Different threads may be creating and deleting the same directory simultaneously, causing some inconsistency issues.

File: azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java (hunk 17 above)
Comment (jamiesjc) [inline azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java]: Thanks! This is very helpful. Seems much easier to implement. I'll give it a try.

File: azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java (hunk 21 above)
Comment (jamiesjc) [inline azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java]: For different tests, the project names and descriptions are different. Might not be easy to combine those together?

File: azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java (hunk 23 above)
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java]: Alternatively, you can consider this method:
http://stackoverflow.com/questions/2593154/get-a-resource-using-getresource

File: azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java (hunk 24 above)
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java]: Got it. Make sense. Thanks

File: azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java (hunk 14 above)
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/project/JdbcProjectLoader.java]: Understood.

File: azkaban-common/src/main/java/azkaban/project/ProjectManager.java (hunk 1 above)
Comment (HappyRay) [inline azkaban-common/src/main/java/azkaban/project/ProjectManager.java]: Good to know that compiler/IDE will take care of this. Thanks

File: azkaban-common/src/main/java/azkaban/project/ProjectManager.java (hunk 12 above)
Comment (jamiesjc) [inline azkaban-common/src/main/java/azkaban/project/ProjectManager.java]: azkaban.project.JdbcProjectLoader#createNewProject(java.sql.Connection, java.lang.String, java.lang.String, azkaban.user.User) will throw the same exception.

File: azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java (hunk 16 above)
Comment (jamiesjc) [inline azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java]: I'll remove this after using the junit tempdir support.

File: azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java (hunk 23 above)
Comment (jamiesjc) [inline azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java]: Changed.

File: azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java (hunk 18 above)
Comment (jamiesjc) [inline azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java]: Removed.

File: azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java (hunk 19 above)
Comment (jamiesjc) [inline azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java]: Changed.

File: azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java (hunk 20 above)
Comment (jamiesjc) [inline azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java]: done.

File: azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java [hunk 25]
```diff
@@ -3,160 +3,180 @@
...
+    }).when(loader).removeProject(project1, user.getUserId());
 
-  @After
-  public void tearDown() throws IOException {
-    System.out.println("Teardown temp dir");
-    synchronized (this) {
-      if (workingDir != null) {
-        FileUtils.deleteDirectory(workingDir);
-        workingDir = null;
+    doAnswer(new Answer<Void>() {
```
Comment (HappyRay) [inline azkaban-common/src/test/java/azkaban/project/ProjectManagerTest.java]: Consider setting this only in the tests that need it.
